### PR TITLE
Update MPI_3D_EULER_ISMAIL_ROE_ENTROPY_CHECK for other fluxes

### LINKED ISSUE #192: Update MPI_3D_EULER_ISMAIL_ROE_ENTROPY_CHECK for other fluxes
- Update  `MPI_3D_EULER_ISMAIL_ROE_ENTROPY_CHECK` such that we check for entropy conservation/KE conservation/pressure preservation etc. with all two-point fluxes added in PR #111
- Modify `periodic_turbulence.cpp` entropy calculation to integrate entropy function at solution nodes.

### PR DESCRIPTION
**Pull request by cpethrick**
Closes #192.

The test, renamed to `MPI_3D_EULER_ENTROPY_CONSERVING_SPLIT_FORMS_CHECK`, now verifies all split form fluxes (Ismail-Roe, Chandrashekar, Ranocha, and Kennedy-Gruber). As well, the `get_numerical_entropy()` function now integrates at solution nodes.

### CODE REVIEW

**Comment by cpethrick** [comment]
I've attached the preliminary results of this test. IR, CH and Ra behave as expected (conserve entropy). As KG is KE-preserving, the entropy change is nonzero, so I have increased the entropy change tolerance such that the test passes. Can anyone think of  a better way to verify the KG flux? I output the KE change over the time interval and it is quite similar between the three tests. @jbrillon @AlexanderCicchino 
[log.txt](https://github.com/dougshidong/PHiLiP/files/10736892/log.txt)

**Comment by jbrillon** [comment]
> 

Other than checking that entropy is not conserved when using KG flux, I'm not sure, do the authors have a test?

File: src/physics/euler.cpp [hunk 1]
```diff
@@ -295,6 +295,22 @@ ::compute_specific_enthalpy ( const std::array<real,nstate> &conservative_soln,
...
+template <int dim, int nstate, typename real>
+inline real Euler<dim,nstate,real>
+::compute_numerical_entropy_function ( const std::array<real,nstate> &conservative_soln ) const
+{
+    const real pressure = compute_pressure<real>(conservative_soln);
+    const real density = conservative_soln[0];
+
+    real entropy = pressure * pow(density, -gam);
+    if (entropy > 0)    entropy = log( entropy );
+    else                entropy = BIG_NUMBER;
```
Comment (jbrillon) [inline src/physics/euler.cpp:307]: maybe put this in a function like I've done in PR #194

Comment (jbrillon) [reply src/physics/euler.cpp:307]: that way you get a message if its negative

File: tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm [hunk 2]
```diff
@@ -14,20 +14,19 @@ set use_collocated_nodes = true
...
 set use_inverse_mass_on_the_fly = true
-# set use_collocated_nodes = true
 
 # Note: this was added to turn off check_same_coords() -- has no other function when dim!=1
 set use_periodic_bc = true
 
 # numerical fluxes
 set conv_num_flux = two_point_flux
-set two_point_num_flux_type = IR 
+set two_point_num_flux_type = IR  # this is changed by the test
```
Comment (jbrillon) [inline tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm:23]: why even pass one then? can remove this line

**Review by jbrillon — APPROVED** [review]
Very clean PR, thanks for making the review quick and easy, great work!

File: tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm [hunk 3]
```diff
@@ -14,20 +14,19 @@ set use_collocated_nodes = true
...
 
 # ODE solver
 subsection ODE solver
   set ode_output = quiet
   set ode_solver_type = runge_kutta
-  set output_solution_every_x_steps = 10
+  set output_solution_every_x_steps = -1
```
Comment (PranshulThakur) [inline tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm:29]: Can be commented out.

File: src/testing/euler_entropy_conserving_split_forms_check.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+              << "Final:                  " << final_entropy << std::endl 
+              << "Scaled difference:      " << abs((initial_entropy-final_entropy)/initial_entropy) 
+              << std::endl << std::endl;
+        pcout << "Initial kinetic energy: " << std::setprecision(16) << initial_KE << std::endl 
+              << "Final:                  " << final_KE << std::endl
+              << "Scaled difference:      " << abs((initial_KE-final_KE)/initial_KE) 
+              << std::endl << std::endl;
+
+        if (abs((initial_entropy-final_entropy)/initial_entropy) > tol){
+            pcout << "Entropy change is not within allowable tolerance. Test failing." << std::endl;
```
Comment (PranshulThakur) [inline src/testing/euler_entropy_conserving_split_forms_check.cpp:65]: Adding an output of which flux is failing might be helpful.

File: src/flow_solver/flow_solver_cases/periodic_turbulence.cpp [hunk 5]
```diff
@@ -295,80 +295,109 @@ double PeriodicTurbulence<dim, nstate>::get_deviatoric_strain_rate_tensor_based_
...
+    OPERATOR::mapping_shape_functions<dim,2*dim> mapping_basis(1, poly_degree, dg->max_grid_degree);
+    mapping_basis.build_1D_shape_functions_at_grid_nodes(dg->high_order_grid->oneD_fe_system, dg->high_order_grid->oneD_grid_nodes);
+    mapping_basis.build_1D_shape_functions_at_flux_nodes(dg->high_order_grid->oneD_fe_system, dg->oneD_quadrature_collection[poly_degree], dg->oneD_face_quadrature);
+
     std::vector<dealii::types::global_dof_index> dofs_indices (n_dofs_cell);
+    
+    double integrand_numerical_entropy_function=0;
+    double integral_numerical_entropy_function=0;
+    const std::vector<double> &quad_weights = dg->volume_quadrature_collection[poly_degree].get_weights();
+    std::shared_ptr < Physics::Euler<dim, nstate, double > > euler_physics = std::dynamic_pointer_cast<Physics::Euler<dim,dim+2,double>>(PHiLiP::Physics::PhysicsFactory<dim,nstate,double>::create_Physics(dg->all_parameters));
```
Comment (PranshulThakur) [inline src/flow_solver/flow_solver_cases/periodic_turbulence.cpp:324]: Is `euler_physics` being used somewhere?

**Review by PranshulThakur — COMMENTED** [review]
Great PR for checking entropy conservation. I have added a few editorial comments.

File: src/testing/euler_entropy_conserving_split_forms_check.cpp (hunk 4 above)
Comment (PranshulThakur) [reply src/testing/euler_entropy_conserving_split_forms_check.cpp:65]: Oh nvm, it's already displayed at the beginning.

Comment (cpethrick) [reply src/testing/euler_entropy_conserving_split_forms_check.cpp:65]: Thanks for the comment, I figured displaying at the beginning is sufficient!

File: tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm (hunk 3 above)
Comment (cpethrick) [reply tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm:29]: Sure, will fix.

File: tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm (hunk 2 above)
Comment (cpethrick) [reply tests/integration_tests_control_files/euler_entropy_conservation/euler_entropy_conserving_split_forms_check.prm:23]: Removed and replaced with a comment.

File: src/flow_solver/flow_solver_cases/periodic_turbulence.cpp (hunk 5 above)
Comment (cpethrick) [reply src/flow_solver/flow_solver_cases/periodic_turbulence.cpp:324]: Good point, I think that is there because I copied code from elsewhere. I'll remove it as `navier_stokes_physics` is used instead.

File: src/physics/euler.cpp (hunk 1 above)
Comment (cpethrick) [reply src/physics/euler.cpp:307]: That's a good suggestion. I've made that change.

**Comment by cpethrick** [comment]
Regarding tests used by the original authors, they don't have a test using a similar setup to this one. From a very quick scan of the results section (https://doi.org/10.1016/j.jcp.2007.09.020), they use a DHIT test case and mostly analyze the spectra to evaluate their flux. I think the way I've implemented the test is fine for as an implementation test - if this test fails in the future we will know that something has changed in the DG split formulation.

**Review by PranshulThakur — APPROVED** [review]
Thanks for addressing the comments. LGTM!

**Comment by sivanadarajah** [comment]
> Closes #192.
> 
> The test, renamed to `MPI_3D_EULER_ENTROPY_CONSERVING_SPLIT_FORMS_CHECK`, now verifies all split form fluxes (Ismail-Roe, Chandrashekar, Ranocha, and Kennedy-Gruber). As well, the `get_numerical_entropy()` function now integrates at solution nodes.

By solution nodes, does the test depend on a specific set of nodes, i.e. LGL or GL.

**Comment by sivanadarajah** [comment]
> I've attached the preliminary results of this test. IR, CH and Ra behave as expected (conserve entropy). As KG is KE-preserving, the entropy change is nonzero, so I have increased the entropy change tolerance such that the test passes. Can anyone think of a better way to verify the KG flux? I output the KE change over the time interval and it is quite similar between the three tests. @jbrillon @AlexanderCicchino [log.txt](https://github.com/dougshidong/PHiLiP/files/10736892/log.txt)

We should have a separate KE test where KG passes the others fail except for Ranocha.

**Comment by cpethrick** [comment]
> > Closes #192.
> > The test, renamed to `MPI_3D_EULER_ENTROPY_CONSERVING_SPLIT_FORMS_CHECK`, now verifies all split form fluxes (Ismail-Roe, Chandrashekar, Ranocha, and Kennedy-Gruber). As well, the `get_numerical_entropy()` function now integrates at solution nodes.
> 
> By solution nodes, does the test depend on a specific set of nodes, i.e. LGL or GL.

I believe that the `compute_numerical_entropy()` function uses the same grid nodes as `dg` - see lines 316 and 317 of `periodic_turbulence.cpp`. @AlexanderCicchino may be able to give a better explanation, as I closely followed a function he wrote to form this one.

**Comment by cpethrick** [comment]
> > I've attached the preliminary results of this test. IR, CH and Ra behave as expected (conserve entropy). As KG is KE-preserving, the entropy change is nonzero, so I have increased the entropy change tolerance such that the test passes. Can anyone think of a better way to verify the KG flux? I output the KE change over the time interval and it is quite similar between the three tests. @jbrillon @AlexanderCicchino [log.txt](https://github.com/dougshidong/PHiLiP/files/10736892/log.txt)
> 
> We should have a separate KE test where KG passes the others fail except for Ranocha.

That was my initial thought too, but the result of the test is difficult to interpret. I output both KE and numerical entropy in the test currently.  The KE change is as follows:

IR: 0.0001600134107610995
CH: 0.0001602602703665387
Ra: 0.000158731496180356
KG: 0.0001598477358319585

To me, it's not completely clear how to interpret these values as "kinetic energy preserving" or "not kinetic energy preserving". Do you have any suggestions?

**Comment by cpethrick** [comment]
Per discussion with @AlexanderCicchino, the kinetic energy change is expected to be nonzero because of the contribution from pressure work. 

My intention with adding these tests is to verify the entropy-conserving scheme in general - as entropy should remain at machine zero, this test will be a good indicator for unintended changes within DG or Euler physics. 

In the future, we could do one or both of: (1) modify this test to account for pressure work, which would be expected to be at machine zero for collocated nodes; (2) add another test to directly verify the fluxes based on the Tadmor shuffle condition.
